shell/platform/darwin/ios/framework/Source/FlutterMetalLayer.mm

```diff
@@ -0,0 +1,402 @@
+    (id)kIOSurfaceHeight : @(_drawableSize.height),
+    (id)kIOSurfacePixelFormat : @(pixelFormat),
+    (id)kIOSurfaceBytesPerElement : @(bytesPerElement),
+    (id)kIOSurfaceBytesPerRow : @(bytesPerRow),
+    (id)kIOSurfaceAllocSize : @(totalBytes),
+  };
+
+  IOSurfaceRef res = IOSurfaceCreate((CFDictionaryRef)options);
+  if (self.colorspace != nil) {
+    CFStringRef name = CGColorSpaceGetName(self.colorspace);
+    IOSurfaceSetValue(res, CFSTR("IOSurfaceColorSpace"), name);
+  } else {
+    IOSurfaceSetValue(res, CFSTR("IOSurfaceColorSpace"), kCGColorSpaceSRGB);
+  }
+  return (__bridge_transfer IOSurface*)res;
+}
+
+- (FlutterTexture*)nextTexture {
+  @synchronized(self) {
+    if (_totalTextures < 3) {
+      ++_totalTextures;
+      IOSurface* surface = [self createIOSurface];
+      MTLTextureDescriptor* textureDescriptor =
+          [MTLTextureDescriptor texture2DDescriptorWithPixelFormat:_pixelFormat
+                                                             width:_drawableSize.width
+                                                            height:_drawableSize.height
+                                                         mipmapped:NO];
+      if (_framebufferOnly) {
+        textureDescriptor.usage = MTLTextureUsageRenderTarget;
+      } else {
+        textureDescriptor.usage =
+            MTLTextureUsageRenderTarget | MTLTextureUsageShaderRead | MTLTextureUsageShaderWrite;
+      }
+      id<MTLTexture> texture = [self.device newTextureWithDescriptor:textureDescriptor
+                                                           iosurface:(__bridge IOSurfaceRef)surface
+                                                               plane:0];
+      FlutterTexture* flutterTexture = [[FlutterTexture alloc] initWithTexture:texture
+                                                                       surface:surface];
+      return flutterTexture;
+    } else {
+      // Make sure raster thread doesn't have too many drawables in flight.
+      if (_availableTextures.count == 0) {
+        CFTimeInterval start = CACurrentMediaTime();
+        while (_availableTextures.count == 0 && CACurrentMediaTime() - start < 1.0) {
+          usleep(100);
+        }
+        CFTimeInterval elapsed = CACurrentMediaTime() - start;
+        if (_availableTextures.count == 0) {
+          NSLog(@"Waited %f seconds for a drawable, giving up.", elapsed);
+          return nil;
+        } else {
+          NSLog(@"Had to wait %f seconds for a drawable", elapsed);
+        }
+      }
+      // Return first drawable that is not in use or the one that was presented
+      // the longest time ago.
+      FlutterTexture* res = nil;
+      for (FlutterTexture* texture in _availableTextures) {
+        if (!texture.surface.isInUse) {
+          res = texture;
+          break;
+        }
+        if (res == nil || texture.presentedTime < res.presentedTime) {
+          res = texture;
+        }
+      }
+      [_availableTextures removeObject:res];
```

Comment:
do we need to check NPE here?

Reply:
What for? Neither `_availableTextures` nor `res` can be `nil` at that place. Also invoking selectors on `nil` objects is permitted in objc.

Reply:
oh i was talking about `res` - is it possible that it's not set?

Reply:
reading it again - it looks like res will always be set. However, it's possible that the texture is in use (`texture.surface.isInUse`). Will that be fine?